File: packages/libbird/android/src/main/java/dev/flutterbird/ladybird/LadybirdPlugin.java
package dev.flutterbird.ladybird;

import android.graphics.Bitmap;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.PorterDuff;
import android.os.Handler;
import android.os.Looper;
import android.view.Surface;
import android.graphics.Rect;

import android.hardware.HardwareBuffer;
import android.graphics.ColorSpace;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.Map;

import io.flutter.embedding.engine.plugins.FlutterPlugin;
import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.common.MethodChannel.MethodCallHandler;
import io.flutter.plugin.common.MethodChannel.Result;
import io.flutter.view.TextureRegistry;

public final class LadybirdPlugin implements FlutterPlugin, MethodCallHandler {

    static {
        // SDL registers JNI methods for org.libsdl.app.* during JNI_OnLoad.
        // Load engine explicitly from Java so JNI_OnLoad has app classloader context.
        System.loadLibrary("engine");
        System.loadLibrary("ladybird_plugin");
    }

    private final Handler mainHandler = new Handler(Looper.getMainLooper());
    private android.os.HandlerThread renderThread;
    private Handler renderHandler;
    private final Map<Long, TextureContext> activeTextures = new HashMap<>();

    private final android.view.Choreographer.FrameCallback frameCallback = new android.view.Choreographer.FrameCallback() {
        @Override
        public void doFrame(long frameTimeNanos) {
            pumpScheduled = false;
            if (activeTextures.isEmpty()) {
                return;
            }

            pumpExecutions += 1;
            pumpTicks += 1;
            nativeTickLadybird();

            for (TextureContext textureContext : new ArrayList<>(activeTextures.values())) {
                renderLatestFrame(textureContext);
            }

            if (!activeTextures.isEmpty()) {
                schedulePump();
            }
        }
    };

    private MethodChannel channel;
    private TextureRegistry textureRegistry;
    private boolean pumpScheduled = false;
    private long pumpRequests = 0;
    private long pumpExecutions = 0;
    private long pumpTicks = 0;

    @Override
    public void onAttachedToEngine(FlutterPlugin.FlutterPluginBinding flutterPluginBinding) {
        textureRegistry = flutterPluginBinding.getTextureRegistry();
        channel = new MethodChannel(flutterPluginBinding.getBinaryMessenger(), "ladybird");
        channel.setMethodCallHandler(this);

        try {
            LadybirdRuntimeFiles.RuntimeConfiguration runtimeConfiguration = LadybirdRuntimeFiles
                    .prepare(flutterPluginBinding.getApplicationContext());
            nativeConfigureAndroid(
                    runtimeConfiguration.resourceRoot,
                    runtimeConfiguration.userDir,
                    runtimeConfiguration.nativeLibraryDir,
                    runtimeConfiguration.certificatesPath);
        } catch (Exception exception) {
            throw new RuntimeException("Failed to prepare Ladybird Android runtime", exception);
        }

        renderThread = new android.os.HandlerThread("LadybirdRenderThread");
        renderThread.start();
        renderHandler = new Handler(renderThread.getLooper());
    }

    @Override
    public void onMethodCall(MethodCall call, Result result) {
        switch (call.method) {
            case "createTexture":
                createTexture(call, result);
                return;
            case "unregisterTexture":
                unregisterTexture(call, result);
                return;
            case "getTextureDiagnostics":
                getTextureDiagnostics(call, result);
                return;
            default:
                result.notImplemented();
        }
    }

    @Override
    public void onDetachedFromEngine(FlutterPlugin.FlutterPluginBinding binding) {
        android.view.Choreographer.getInstance().removeFrameCallback(frameCallback);
        pumpScheduled = false;

        for (TextureContext textureContext : activeTextures.values()) {
            textureContext.release();
        }
        activeTextures.clear();

        channel.setMethodCallHandler(null);
        textureRegistry = null;

        if (renderThread != null) {
            renderThread.quitSafely();
            renderThread = null;
            renderHandler = null;
        }
    }

    private void createTexture(MethodCall call, Result result) {
        if (!(call.arguments instanceof Number) || textureRegistry == null) {
            result.error("INVALID_ARGS", "Expected numeric view ID", null);
            return;
        }

        int viewId = ((Number) call.arguments).intValue();
        TextureRegistry.SurfaceProducer producer = textureRegistry.createSurfaceProducer();
        TextureContext textureContext = new TextureContext(viewId, producer);
        producer.setCallback(new TextureRegistry.SurfaceProducer.Callback() {
            @Override
            public void onSurfaceDestroyed() {
            }

            @Override
            public void onSurfaceAvailable() {
                schedulePump();
            }
        });

        activeTextures.put(textureContext.textureId, textureContext);
        renderLatestFrame(textureContext);
        updatePumpDriver();
        result.success(textureContext.textureId);
    }

    private void unregisterTexture(MethodCall call, Result result) {
        if (!(call.arguments instanceof Number)) {
            result.error("INVALID_ARGS", "Expected numeric texture ID", null);
            return;
        }

        long textureId = ((Number) call.arguments).longValue();
        TextureContext textureContext = activeTextures.remove(textureId);
        if (textureContext != null) {
            textureContext.release();
        }

        updatePumpDriver();
        result.success(null);
    }

    private void getTextureDiagnostics(MethodCall call, Result result) {
        if (!(call.arguments instanceof Number)) {
            result.error("INVALID_ARGS", "Expected numeric texture ID", null);
            return;
        }

        long textureId = ((Number) call.arguments).longValue();
        TextureContext textureContext = activeTextures.get(textureId);
        if (textureContext == null) {
            result.success(null);
            return;
        }

        Map<String, Object> diagnostics = new HashMap<>();
        diagnostics.put("textureId", textureContext.textureId);
        diagnostics.put("viewId", textureContext.viewId);
        diagnostics.put("isActive", textureContext.active);
        diagnostics.put("frameNotifyQueued", textureContext.frameNotifyQueued);
        diagnostics.put("queuedGeneration", textureContext.lastFrameGeneration);
        diagnostics.put("lastFrameGeneration", textureContext.lastFrameGeneration);
        diagnostics.put("nativeFrameCallbacks", textureContext.nativeFrameCallbacks);
        diagnostics.put("queuedDrops", textureContext.queuedDrops);
        diagnostics.put("deliveredFrames", textureContext.deliveredFrames);
        diagnostics.put("displayLinkTicks", pumpTicks);
        diagnostics.put("pumpRequests", pumpRequests);
        diagnostics.put("pumpExecutions", pumpExecutions);
        diagnostics.put("hasDisplayLink", true);
        result.success(diagnostics);
    }

    private void renderLatestFrame(TextureContext textureContext) {
        if (!textureContext.active) {
            return;
        }

        long generation = nativeGetFrameGeneration(textureContext.viewId);
        textureContext.nativeFrameCallbacks += 1;
        if (generation == 0 || generation == textureContext.lastFrameGeneration) {
            return;
        }

        if (textureContext.frameNotifyQueued) {
            return;
        }

        if (textureContext.lastFrameGeneration > 0 && generation > textureContext.lastFrameGeneration + 1) {
            textureContext.queuedDrops += (generation - textureContext.lastFrameGeneration - 1);
        }

        int width = nativeGetSurfaceWidth(textureContext.viewId);
        int height = nativeGetSurfaceHeight(textureContext.viewId);
        if (width <= 0 || height <= 0) {
            return;
        }

        textureContext.ensureFrameStorage(width, height);

        final HardwareBuffer buffer = nativeGetHardwareBuffer(textureContext.viewId);
        if (buffer == null) {
            android.util.Log.e("LadybirdPlugin", "nativeGetHardwareBuffer returned null!");
            textureContext.queuedDrops += 1;
            return;
        }

        final Surface surface = textureContext.producer.getSurface();
        if (surface == null || !surface.isValid()) {
            android.util.Log.e("LadybirdPlugin", "Surface is null or invalid!");
            buffer.close();
            textureContext.queuedDrops += 1;
            return;
        }

        textureContext.lastFrameGeneration = generation;
        textureContext.frameNotifyQueued = true;

        if (renderHandler != null) {
            renderHandler.post(new Runnable() {
                @Override
                public void run() {
                    Canvas canvas = null;
                    try {
                        canvas = surface.lockHardwareCanvas();
                        canvas.drawColor(Color.TRANSPARENT, PorterDuff.Mode.CLEAR);
                        Bitmap bitmap = Bitmap.wrapHardwareBuffer(buffer, ColorSpace.get(ColorSpace.Named.SRGB));
                        if (bitmap != null) {
                            Rect src = new Rect(0, 0, width, height);
                            Rect dst = new Rect(0, 0, width, height);
                            canvas.drawBitmap(bitmap, src, dst, null);
                        } else {
                            android.util.Log.e("LadybirdPlugin", "Bitmap.wrapHardwareBuffer returned null!");
                        }
                        textureContext.deliveredFrames += 1;
                    } catch (RuntimeException exception) {
                        android.util.Log.e("LadybirdPlugin", "Exception drawing hardware buffer", exception);
                        textureContext.queuedDrops += 1;
                    } finally {
                        if (canvas != null) {
                            surface.unlockCanvasAndPost(canvas);
                        }
                        buffer.close();
                        textureContext.frameNotifyQueued = false;
                    }
                }
            });
        } else {
            buffer.close();
            textureContext.frameNotifyQueued = false;
        }
    }

    private void updatePumpDriver() {
        if (activeTextures.isEmpty()) {
            android.view.Choreographer.getInstance().removeFrameCallback(frameCallback);
            pumpScheduled = false;
            return;
        }

        schedulePump();
    }

    private void schedulePump() {
        if (pumpScheduled) {
            return;
        }

        pumpScheduled = true;
        pumpRequests += 1;
        android.view.Choreographer.getInstance().postFrameCallback(frameCallback);
    }

    private static final class TextureContext {
        final int viewId;
        final TextureRegistry.SurfaceProducer producer;
        final long textureId;

        boolean active = true;
        volatile boolean frameNotifyQueued = false;
        int width = 0;
        int height = 0;
        long lastFrameGeneration = 0;
        long nativeFrameCallbacks = 0;
        long queuedDrops = 0;
        long deliveredFrames = 0;

        TextureContext(int viewId, TextureRegistry.SurfaceProducer producer) {
            this.viewId = viewId;
            this.producer = producer;
            this.textureId = producer.id();
        }

        void ensureFrameStorage(int nextWidth, int nextHeight) {
            if (nextWidth == width && nextHeight == height) {
                return;
            }

            width = nextWidth;
            height = nextHeight;
            producer.setSize(width, height);
        }

        void release() {
            active = false;
            producer.release();
        }
    }

    private static native void nativeConfigureAndroid(
            String resourceRoot,
            String userDir,
            String nativeLibraryDir,
            String certificatesPath);

    private static native void nativeTickLadybird();

    private static native long nativeGetFrameGeneration(int viewId);

    private static native int nativeGetSurfaceWidth(int viewId);

    private static native int nativeGetSurfaceHeight(int viewId);

    private static native HardwareBuffer nativeGetHardwareBuffer(int viewId);
}

File: packages/libbird/android/src/main/kotlin/com/example/ladybird/LadybirdPlugin.kt
package com.example.ladybird

import android.content.Context
import android.content.Intent
import android.content.ServiceConnection
import android.graphics.Bitmap
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.PorterDuff
import android.util.Log
import android.view.Choreographer
import android.view.Surface
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.FlutterPlugin.FlutterPluginBinding
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import io.flutter.view.TextureRegistry
import java.nio.ByteBuffer
import org.serenityos.ladybird.LadybirdServiceConnection
import org.serenityos.ladybird.WebContentService

class LadybirdPlugin : FlutterPlugin, MethodCallHandler, Choreographer.FrameCallback {
  private lateinit var channel: MethodChannel
  private var binding: FlutterPluginBinding? = null
  private var textureRegistry: TextureRegistry? = null
  private val textures = mutableMapOf<Long, TextureState>()
  private val serviceConnections = mutableListOf<ServiceConnection>()
  private var attached = false

  private data class TextureState(
    val viewId: Int,
    val entry: TextureRegistry.SurfaceTextureEntry,
    val surface: Surface,
    var bitmap: Bitmap? = null,
  )

  override fun onAttachedToEngine(binding: FlutterPluginBinding) {
    this.binding = binding
    System.loadLibrary("ladybird_jni")
    nativeInitLadybird()

    textureRegistry = binding.textureRegistry
    channel = MethodChannel(binding.binaryMessenger, "ladybird")
    channel.setMethodCallHandler(this)
    attached = true
    Choreographer.getInstance().postFrameCallback(this)
  }

  override fun onDetachedFromEngine(binding: FlutterPluginBinding) {
    attached = false
    Choreographer.getInstance().removeFrameCallback(this)
    val appContext = this.binding?.applicationContext
    if (appContext != null) {
      for (connection in serviceConnections) {
        runCatching { appContext.unbindService(connection) }
      }
    }
    serviceConnections.clear()
    releaseAllTextures()
    channel.setMethodCallHandler(null)
    textureRegistry = null
    this.binding = null
  }

  fun bindWebContentServiceFromNative(ipcFd: Int) {
    val pluginBinding = binding ?: return
    val appContext = pluginBinding.applicationContext
    val resourceDir = appContext.dataDir.absolutePath

    val connector = LadybirdServiceConnection(ipcFd, resourceDir)
    connector.onDisconnect = {
      Log.e("LadybirdPlugin", "WebContent service disconnected")
    }

    val bound = appContext.bindService(
      Intent(appContext, WebContentService::class.java),
      connector,
      Context.BIND_AUTO_CREATE,
    )

    if (bound) {
      serviceConnections.add(connector)
    } else {
      Log.e("LadybirdPlugin", "Failed to bind WebContentService")
    }
  }

  override fun onMethodCall(call: MethodCall, result: Result) {
    when (call.method) {
      "createTexture" -> {
        val viewId = (call.arguments as? Number)?.toInt()
        if (viewId == null) {
          result.error("INVALID_ARGS", "Expected int viewId", null)
          return
        }

        val registry = textureRegistry
        if (registry == null) {
          result.error("UNAVAILABLE", "Texture registry is null", null)
          return
        }

        val entry = registry.createSurfaceTexture()
        val surface = Surface(entry.surfaceTexture())
        val textureId = entry.id()

        textures[textureId]?.let { existing ->
          existing.surface.release()
          existing.entry.release()
        }

        textures[textureId] = TextureState(
          viewId = viewId,
          entry = entry,
          surface = surface,
        )

        result.success(textureId)
      }

      "unregisterTexture" -> {
        val textureId = (call.arguments as? Number)?.toLong()
        if (textureId == null) {
          result.error("INVALID_ARGS", "Expected int textureId", null)
          return
        }

        releaseTexture(textureId)
        result.success(null)
      }

      else -> result.notImplemented()
    }
  }

  override fun doFrame(frameTimeNanos: Long) {
    if (!attached) {
      return
    }

    if (textures.isNotEmpty()) {
      nativeTickLadybird()
    }

    // Draw all active views to their bound surfaces once per frame.
    for (state in textures.values) {
      renderTexture(state)
    }

    Choreographer.getInstance().postFrameCallback(this)
  }

  private fun renderTexture(state: TextureState) {
    val width = nativeGetTextureWidth(state.viewId)
    val height = nativeGetTextureHeight(state.viewId)
    if (width <= 0 || height <= 0) {
      return
    }

    val byteBuffer = nativeGetLatestPixelBuffer(state.viewId)
    if (byteBuffer == null || byteBuffer.capacity() < width * height * 4) {
      return
    }

    state.entry.surfaceTexture().setDefaultBufferSize(width, height)

    var bitmap = state.bitmap
    if (bitmap == null || bitmap.width != width || bitmap.height != height || bitmap.isRecycled) {
      bitmap?.recycle()
      bitmap = Bitmap.createBitmap(width, height, Bitmap.Config.ARGB_8888)
      state.bitmap = bitmap
    }

    byteBuffer.rewind()
    bitmap.copyPixelsFromBuffer(byteBuffer)

    var canvas: Canvas? = null
    try {
      canvas = state.surface.lockCanvas(null)
      canvas.drawColor(Color.TRANSPARENT, PorterDuff.Mode.SRC)
      canvas.drawBitmap(bitmap, 0f, 0f, null)
    } catch (_: Throwable) {
      // Ignore frame draw failures and retry on next frame.
    } finally {
      if (canvas != null) {
        state.surface.unlockCanvasAndPost(canvas)
      }
    }
  }

  private fun releaseTexture(textureId: Long) {
    val state = textures.remove(textureId) ?: return
    state.bitmap?.recycle()
    state.surface.release()
    state.entry.release()
  }

  private fun releaseAllTextures() {
    val ids = textures.keys.toList()
    for (id in ids) {
      releaseTexture(id)
    }
  }

  private external fun nativeTickLadybird()

  private external fun nativeInitLadybird()

  private external fun nativeGetLatestPixelBuffer(viewId: Int): ByteBuffer?

  private external fun nativeGetTextureWidth(viewId: Int): Int

  private external fun nativeGetTextureHeight(viewId: Int): Int
}
